core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java

```diff
@@ -2107,14 +2333,28 @@ private void deleteOneFromRemote(
       final MongoNamespace namespace,
       final BsonValue documentId
   ) {
-    // TODO: lock down id
-    final CoreDocumentSynchronizationConfig config =
-        syncConfig.getSynchronizedDocument(namespace, documentId);
-    if (config == null) {
-      return;
+    final Lock lock =
+        this.syncConfig.getNamespaceConfig(namespace).getLock().writeLock();
+    lock.lock();
+    try {
+      final CoreDocumentSynchronizationConfig config =
+          syncConfig.getSynchronizedDocument(namespace, documentId);
+      if (config == null) {
+        return;
+      }
+      final BsonDocument documentToDelete = getLocalCollection(namespace)
+          .find(getDocumentIdFilter(documentId)).first();
+      if (documentToDelete == null) {
+        desyncDocumentFromRemote(namespace, documentId);
+        return;
+      }
+      getUndoCollection(namespace).insertOne(documentToDelete);
```

Comment:
[nit] This applies throughout this file, but why not have a `final MongoCollection<..> undoColl = getUndoCollection(...)` or `final MongoCollection<...> localColl = getlocalColl(...)` in places you use the utility functions so you don't unnecessarily create the collection object from scratch every time you do an operation?

Reply:
laziness. will address